libslide/src/scanner.rs

```diff
@@ -1,18 +1,28 @@
 pub mod types;
 
+use crate::diagnostics::Diagnostic;
 use types::TokenType as TT;
-use types::*;
+pub use types::*;
 
-pub fn scan<T: Into<String>>(input: T) -> Vec<Token> {
-    let mut scanner = Scanner::new(input);
+pub struct ScanResult {
+    pub tokens: Vec<Token>,
+    pub diagnostics: Vec<Diagnostic>,
+}
+
+pub fn scan<'a, T: Into<&'a str>>(input: T) -> ScanResult {
+    let mut scanner = Scanner::new(input.into());
     scanner.scan();
-    scanner.output
+    ScanResult {
+        tokens: scanner.output,
+        diagnostics: scanner.diagnostics,
+    }
 }
```

Comment:
Idk about this interface, it's not exactly the most "rusty" interface but I can't figure out a better way without exposing the Scanner object, which I don't think is either necessary or desirable. LMK

Reply:
I think this is a fine way of handling it. Maybe this could be an option wrapped. I'm not sure.